pymtl3/stdlib/ifcs/CalleeIfc.py

```diff
@@ -0,0 +1,167 @@
+        s.ifc_rtl_caller.args = s.args
+        s.called = False
+
+      # add constraints between callee method and upblk
+      s.add_constraints( M( s.cl_callee_method ) < U( update_en_args ) )
+
+      # know which method to add constraints on later
+      cl_method = s.cl_callee_method
+
+    else:
+      # no args, update en
+      @s.update
+      def update_en():
+        s.ifc_rtl_caller.en = Bits1( 1 ) if s.called else Bits1( 0 )
+        s.called = False
+
+      # add constraints between callee method and upblk
+      s.add_constraints( M( s.cl_callee_method_no_arg ) < U( update_en ) )
+
+      # know which method to add constraints on later
+      cl_method = s.cl_callee_method_no_arg
+
+    # generate upblk depending on rets
+    if s.ifc_rtl_caller.rets:
+      # create tmp var for rets
+      s.rets = s.ifc_rtl_caller.rets._dsl.Type()
+
+      # generate upblk for rets
+      @s.update
+      def update_rets():
+        s.rets = s.ifc_rtl_caller.rets
+
+      s.add_constraints( U( update_rets ) < M( cl_method ) )
+
+    else:
+      # return None if no ret
+      s.rets = None
+
+    # Generate upblk and add constraints for rdy
+    @s.update
+    def update_rdy():
+      s.rdy = True if s.ifc_rtl_caller.rdy else False
+
+    s.add_constraints( U( update_rdy ) < M( cl_method ) )
+
+  @non_blocking( lambda s: s.rdy )
+  def cl_callee_method( s, **kwargs ):
+    s.args.__dict__.update( kwargs )
+    s.called = True
+    return s.rets
+
+  @non_blocking( lambda s: s.rdy )
+  def cl_callee_method_no_arg( s ):
+    s.called = True
+    return s.rets
+
+
+#-------------------------------------------------------------------------
+# CalleeIfcRTL
+#-------------------------------------------------------------------------
+class CalleeIfcRTL( Interface ):
+
+  def construct( s, ArgTypes=None, RetTypes=None ):
+    if ArgTypes:
+      # mangle arg bit_struct name by fields
+      arg_cls_name = "CalleeIfcRTL_Arg"
+      for arg_name, arg_type in ArgTypes:
+        arg_cls_name += "_{}_{}".format( arg_name, arg_type.__name__ )
+
+      s.args = InPort( mk_bit_struct( arg_cls_name, ArgTypes ) )
+
+    else:
+      s.args = None
+
+    if RetTypes:
+      # mangle ret bit_struct name by fields
+      ret_cls_name = "CalleeIfcRTL_Ret"
+      for ret_name, ret_type in RetTypes:
+        ret_cls_name += "__{}_{}".format( ret_name, ret_type.__name__ )
+
+      s.rets = OutPort( mk_bit_struct( ret_cls_name, RetTypes ) )
+
+    else:
+      s.rets = None
+
+    s.en = InPort( Bits1 )
+    s.rdy = OutPort( Bits1 )
+
+  def connect( s, other, parent ):
+
+    def connect_callee_ifc( this, other ):
+      if this.args:
+        assert other.args
+        parent.connect( other.args, this.args )
+      if this.rets:
+        assert other.rets
+        parent.connect( this.rets, other.rets )
+
+      parent.connect_pairs( this.en, other.en, this.rdy, other.rdy )
+
+    if isinstance( other, CalleeIfcRTL ):
```

Comment:
If you want to do connect_by_name, you don't need to do anything here. Return False will revert to the default scheme.

Reply:
fixed -> https://github.com/cornell-brg/pymtl3/blob/6db73d99755ce8c11ed2b97b30751cadec0f76c2/pymtl3/stdlib/ifcs/generic_method_ifcs.py#L145-L165